File: trees/insert_delete_bst.cpp
#include <iostream>
#include <queue>

using namespace std;

class Node {
    public:
        int data;
        Node* left;
        Node* right;

        Node(int data) {
            this->data = data;
            this->left = NULL;
            this->right = NULL;
        }
};

class Tree {
    public:
        Node* root;

        Tree() {
            this->root = NULL;
        }

        void destroyTree() {
            if(root == NULL) {
                return;
            }

            Node* root = this->root;
            this->destroyTreeUtil(root);
            root = NULL; 
        }

        void printInorder() {
            Node* root = this->root;
            if(this->root == NULL) {
                cout << "empty root" << endl;
            }
            this->printInorderUtil(root);
            cout << endl;
            return;
        }

        void printInorderUtil(Node* root) {
            if(root == NULL) {
                return;
            }

            printInorderUtil(root->left);
            cout << root->data << " ";
            printInorderUtil(root->right);
        }

        void insertBST(int data) {
            if(this->root == NULL) {
                this->root = new Node(data);
                return;
            }

            // root has some data
            Node* root = this->root;
            this->insertBSTUtil(root, data);
            return;
        }

        void insertLowestLevel(int data) {
            if(this->root == NULL) {
                this->root = new Node(data);
                return;
            }

            // root has some data
            Node* root = this->root;
            this->insertLowestLevelUtil(root, data);
            return;
        }

        void deleteBST(int data) {
            if(this->root == NULL) {
                return;
            }

            // 3 cases
            //  Node to be deleted is leaf
            //  Node to be deleted is only one child
            //  Node to be deleted has two children
            this->deleteBSTUtil(root, data);
            return;
        }

        Node* deleteBSTUtil(Node* root, int data) {
            if(root == NULL) {
                return NULL;
            }

            // if root is to be deleted
            if(root->data == data) {
                Node* in_order_successor = getInorderSuccessor(root->right);
                root->data = in_order_successor->data;
                root->right = deleteBSTUtil(root->right, in_order_successor->data);
                return root;
            }

            // If left node is the node to be deleted
            Node* temp = root->left;
            if(temp && (temp->data == data)) {
                // If no child
                if(temp->left == NULL && temp->right == NULL) {
                    delete temp;
                    root->left = NULL;
                    return root;
                // if no left child, delete root and return right child
                } else if(temp->left == NULL) {
                    root->left = temp->right;
                    temp->right = NULL;
                    delete(temp);
                    return root;
                // if no right child, delete root and return left child
                } else if(temp->right == NULL) {
                    root->left = temp->left;
                    temp->left = NULL;
                    delete(temp);
                    return root;
                // Both child
                } else {
                    Node* in_order_successor = getInorderSuccessor(temp->right);
                    temp->data = in_order_successor->data;
                    temp->right = deleteBSTUtil(root->right, temp->data);
                    return root;
                }
            }

            // If right node is the node to be deleted
            temp = root->right;
            if(temp && (temp->data == data)) {
                // If no child
                if(temp->left == NULL && temp->right == NULL) {
                    delete temp;
                    root->right = NULL;
                    return root;
                // if no left child, delete root and return right child
                } else if(temp->left == NULL) {
                    root->right = temp->right;
                    temp->right = NULL;
                    delete(temp);
                    return root;
                // if no right child, delete root and return left child
                } else if(temp->right == NULL) {
                    root->right = temp->left;
                    temp->left = NULL;
                    delete(temp);
                    return root;
                // Both child
                } else {
                    Node* in_order_successor = getInorderSuccessor(temp->right);
                    temp->data = in_order_successor->data;
                    temp->right = deleteBSTUtil(root->right, temp->data);
                    return root;
                }
            }

            if(data < root->data) {
                root->left = deleteBSTUtil(root->left, data);
            } else if(data > root->data) {
                root->right = deleteBSTUtil(root->right, data);
            }

            return root;
        }

        void totalNodes() {
            if(this->root == NULL) {
                return;
            }

            Node* root = this->root;
            cout << "Total nodes in tree: " << totalNodesUtil(root) << endl;
            return;
        }

        void getMaximumWidth() {
            if(this->root == NULL) {
                return;
            }

            Node* root = this->root;
            int height = getHeight(root);
            vector<int> widthCount;
            for(int i = 0; i < height; i++) {
                widthCount.push_back(0);
            }
            getMaximumWidthHelper(root, widthCount, 0);

            int maxWidth = 0;
            for(int i = 0; i < widthCount.size(); i++) {
                if(widthCount[i] > maxWidth) {
                     maxWidth = widthCount[i];
                }
            }

            cout << "Max width recursively: " << maxWidth << endl;
            return;
        }

        void diameter() {
            if(this->root == NULL) {
                return;
            }

            Node* root = this->root;

            int sum = 0;
            diameterUtil(root, sum);
            cout << "Diameter: " << sum << endl;
            return;
        }

    private:
        void insertBSTUtil(Node* root, int data) {
            if(root == NULL) {
                root = new Node(data);
                return;
            }

            if((data < root->data) && (root->left == NULL)) {
                root->left = new Node(data);
                return;
            } else if((data > root->data) && (root->right == NULL)) {
                root->right = new Node(data);
                return;
            }

            if(data < root->data) {
                insertBSTUtil(root->left, data);
            } else {
                insertBSTUtil(root->right, data);
            }

            return;
        }

        void insertLowestLevelUtil(Node* root, int data) {
            if(root == NULL) {
                root = new Node(data);
                return;
            }

            queue<Node*> level_order_queue;
            level_order_queue.push(root);

            while(!level_order_queue.empty()) {
                Node* front = level_order_queue.front();
                level_order_queue.pop();

                if(front->left == NULL) {
                    front->left = new Node(data);
                    return;
                } else if(front->right == NULL) {
                    front->right = new Node(data);
                    return;
                }

                level_order_queue.push(front->left);
                level_order_queue.push(front->right);
            }
            return;
        }

        void destroyTreeUtil(Node* root) {
            if(root == NULL) {
                return;
            }

            destroyTreeUtil(root->right);
            destroyTreeUtil(root->left);
            root->left = NULL;
            root->right = NULL;
            delete root;
        }

        Node* getInorderSuccessor(Node* root) {
            if(root == NULL) {
                return NULL;
            }

            while(root->left != NULL) {
                root = root->left;
            }
            return root; 
        }

        int totalNodesUtil(Node* root) {
            if(root == NULL) {
                return 0;
            }

            return 1 + totalNodesUtil(root->left) + totalNodesUtil(root->right);
        }

        int getHeight(Node* root) {
            if(root == NULL) {
                return 0;
            }

            int lHeight = getHeight(root->left);
            int rHeight = getHeight(root->right);

            return max(lHeight, rHeight) + 1;
        }

        void  getMaximumWidthHelper(Node* root, vector<int> &widthCount, int level) {
            if(root == NULL) {
                return;
            }

            widthCount[level]++;
            getMaximumWidthHelper(root->left, widthCount, level+1);
            getMaximumWidthHelper(root->right, widthCount, level+1);
        }

        int diameterUtil(Node* root, int &sum) {
            if(root == NULL) {
                return 0;
            }

            int left = diameterUtil(root->left, sum);
            int right = diameterUtil(root->right, sum);

            if(left + right + 1 > sum) {
                sum = left + right + 1;
            }
            return std::max(left, right) + 1;
        }
};

int main() {

    Tree tree;

    tree.insertBST(50);
    tree.insertBST(30);
    tree.insertBST(20);
    tree.insertBST(40);
    tree.insertBST(70);
    tree.insertBST(80);
    tree.printInorder();
    tree.totalNodes();

    // insert a node at lowerst possible level in the tree
    tree.insertLowestLevel(60);
    tree.printInorder();

    // delete a node
    tree.deleteBST(50);
    tree.printInorder();

    // O(N) time and O(N) space
    tree.getMaximumWidth();

    // O(N) time complexity
    tree.diameter();

    // free memory
    tree.destroyTree();

    return 0;
}

File: trees/trees_py/insert_delete_BST.py
"""
Find the lowest common ancestor of 2 tree nodes
"""

# -*- encoding: utf-8 *-*


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class Tree:
    def __init__(self):
        self.root = None

    def print_level_order_traversal(self, root):
        if root is None:
            return None

        queue = []
        queue.append(root)
        print()
        while(len(queue) > 0):
            count = len(queue)
            while(count > 0):
                curr_node = queue.pop(0)
                print(curr_node.data, end=" ")
                if curr_node.left:
                    queue.append(curr_node.left)
                if curr_node.right:
                    queue.append(curr_node.right)
                count -= 1
            print()

    def in_order(self, root):
        if root is None:
            return
        self.in_order(root.left)
        print(root.data, end=" ")
        self.in_order(root.right)

    def insert_BST(self, root, val):
        if root is None:
            root = Node(val)
            return root

        if val > root.data and root.right is None:
            root.right = Node(val)
            return root
        elif val < root.data and root.left is None:
            root.left = Node(val)
            return root

        if val < root.data:
            self.insert_BST(root.left, val)
        else:
            self.insert_BST(root.right, val)

        return root

    def insert_lowest_level(self, root, val):
        if root is None:
            root = Node(val)
            return root

        queue = []
        queue.append(root)
        while(len(queue) > 0):
            curr_node = queue.pop(0)
            if curr_node.left is None:
                curr_node.left = Node(val)
                return root
            elif curr_node.right is None:
                curr_node.right = Node(val)
                return root
            queue.append(root.left)
            queue.append(root.right)
        return root

    def get_in_order_successor(self, root):
        """
        Give the node for which to find in_order successor
        First find in the right
        """
        if root is None:
            return None

        while(root.left is not None):
            root = root.left
        return root

    def delete_node_util(self, root, val):
        """
            Root is not the node to be deleted

            # - find node to be deleted in left subtree with a front runner
            # - find node to be deleted in right subtree with a front runner
            # - on finding the node to deleted, there are 4 cases:
            #     - Node with no children:
            #         - Can be deleted without any problem
            #     - Node with both children:
            #         - Find the minimum node in the right subtree (in_order successor)
            #         - Swap the data of root and in_order successor.
            #         - The node to be deleted is now a leaf node
            #         - Find and delete this leaf node
            #     - Node with left child
            #         Remove the node and join parent and child of node to be deleted
            #     - Node with right child
            #         Remove the node and join parent and child of node to be deleted
        """
        temp_node = root.left
        if temp_node is not None and temp_node.data == val:
            if temp_node.left is None and temp_node.right is None:
                temp_node = None
                root.left = None
            elif temp_node.right is not None:
                in_order_successor = self.get_in_order_successor(temp_node.right)
                temp_node.data = in_order_successor.data
                in_order_successor.data = val
                self.delete_node_util(temp_node, val)
            else:
                in_order_successor = self.get_in_order_successor(temp_node.left)
                temp_node.data = in_order_successor.data
                in_order_successor.data = val
                self.delete_node_util(temp_node, val)
            return root

        temp_node = root.right
        if temp_node is not None and temp_node.data == val:
            if temp_node.left is None and temp_node.right is None:
                temp_node = None
                root.right = None
            elif temp_node.right is not None:
                in_order_successor = self.get_in_order_successor(temp_node.right)
                temp_node.data = in_order_successor.data
                in_order_successor.data = val
                self.delete_node_util(temp_node, val)
            else:
                in_order_successor = self.get_in_order_successor(temp_node.left)
                temp_node.data = in_order_successor.data
                in_order_successor.data = val
                self.delete_node_util(temp_node, val)
            return root

        if val < root.data:
            root.left = self.delete_node_util(root.left, val)
        elif val > root.data:
            root.right = self.delete_node_util(root.right, val)

        return root

    def delete_node(self, root, val):
        """
        Deleting a node from a BST

        Case 1: root is the node to be deleted
            Case 1: Root has no children
                make root None and return
            Case 2: Root has right subtree
                - Find in_order successor in right subtree, make root val as in_order successor
                - case 2 in right subtree of root
            Case 3: Root has left subtree
                - Find in_order successor in left subtree, make root val as in_order successor
                - case 2 in left subtree of root

        Case 2: root is not the node to be deleted
            - find node to be deleted in left subtree with a front runner
            - find node to be deleted in right subtree with a front runner
            - on finding the node to deleted, there are 4 cases:
                - Node with no children:
                    - Can be deleted without any problem
                - Node with both children:
                    - Find the minimum node in the right subtree (in_order successor)
                    - Swap the data of root and in_order successor.
                    - The node to be deleted is now a leaf node
                    - Find and delete this leaf node
                - Node with left child
                    Remove the node and join parent and child of node to be deleted
                - Node with right child
                    Remove the node and join parent and child of node to be deleted
        """
        if root is None:
            return

        # If root is the node to be deleted
        if root.data == val:
            if root.left is None and root.right is None:
                root = None
                return root
            elif root.right is not None:
                in_order_successor = self.get_in_order_successor(root.right)
                root.data = in_order_successor.data
                root.right = self.delete_node_util(root.right, in_order_successor.data)
                return root
            else:
                in_order_successor = self.get_in_order_successor(root.left)
                root.data = in_order_successor.data
                root.left = self.delete_node_util(root.leaf, in_order_successor.data)
                return root
        else:
            return self.delete_node_util(root, val)


def main():
    tree = Tree()
    tree.root = tree.insert_BST(tree.root, 50)
    tree.insert_BST(tree.root, 30)
    tree.insert_BST(tree.root, 20)
    tree.insert_BST(tree.root, 40)
    tree.insert_BST(tree.root, 70)
    tree.insert_BST(tree.root, 80)

    tree.in_order(tree.root)
    print()

    # insert a node at lowerst possible level in the tree
    tree.insert_lowest_level(tree.root, 60)

    tree.print_level_order_traversal(tree.root)
    print()

    # Delete a node from BST
    val = 60
    tree.delete_node(tree.root, val)
    tree.print_level_order_traversal(tree.root)
    print()
    return


if __name__ == '__main__':
    main()
